### PR TITLE
Big PR

### PR DESCRIPTION
**Pull request by romgrk**
I had some time to spare today and I ended up doing quite a few modifications to this plugin. I realize this is going to be a big PR, but it still contains substantial improvements so I'd feel bad to keep it for myself. I don't insist on it being merged, I will maintain the differences in my fork if it's too substantial.

The main changes are:
 - Support coc.nvim in the builtin diagnostics
 - Include text/hl groups manipulation functions  (extracted from [barbar](https://github.com/romgrk/barbar.nvim)
 - Use them to provide proper layout & truncation in `loclist.lua`. You can observe the result in the `TODOs` section of the screenshot
 - Add/fix/modify HL groups to allow for more customization

And I still plan to do some more work.

image

### CODE REVIEW

**Comment by GustavoKatel** [comment]
thanks a lot for the PR. this looks very cool. I didn't have much time to review this yet, but I will try to get to it in a few days.

**Comment by romgrk** [comment]
No worries, take your time, I have a few repos I maintain myself, totally get the lack of time to review everything :| If you have questions about why things are done in a way, leave `?` marks I'll comment on them. I've been using this PR for a week, no bug to report until now.

**Review by GustavoKatel — COMMENTED** [review]
this is a really good set of changes! thanks a lot! I only have a couple of questions

File: lua/sidebar-nvim/builtin/diagnostics.lua [hunk 1]
```diff
@@ -74,21 +117,34 @@ local function get_diagnostics()
...
+                augroup END
+            ]], false)
+        else
+            vim.api.nvim_exec([[
+                augroup sidebar_nvim_diagnostics_update
+                    autocmd!
+                    autocmd DiagnosticChanged * lua require'sidebar-nvim.builtin.diagnostics'.update()
+                augroup END
+            ]], false)
+            get_diagnostics()
```
[the remark this answers is not in the document]
Comment (GustavoKatel) [reply lua/sidebar-nvim/builtin/diagnostics.lua:146]: 🤔 `setup` is called only once and we still call multiple times in the `update`.

You're saying that calling `CocAction('diagnosticList')` when there's no event from `CocDiagnosticChange` causes issues with coc?

File: lua/sidebar-nvim/builtin/diagnostics.lua [hunk 2]
```diff
@@ -117,18 +173,22 @@ return {
         },
     },
     bindings = {
-        ["t"] = function(line)
-            loclist:toggle_group_at(line)
+        ["r"] = function()
+            loclist:clear({ remove_groups = true })
         end,
-        ["e"] = function(line)
```
[the remark this answers is not in the document]
Comment (GustavoKatel) [reply lua/sidebar-nvim/builtin/diagnostics.lua:123]: I think we can have:

key | action
----- | ----
`t` | toggle only
`r` | clear
`e` | edit only
`o` | open or toggle

we all win :)

**Comment by romgrk** [comment]
Closing, stale, haven't had time for it :|

**Pull request closed without merging**